src/commonTest/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWalletIntegrationTest.kt

```diff
@@ -0,0 +1,95 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import kotlin.test.Test
+import kotlin.test.assertContains
+import kotlin.test.assertEquals
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+
+@OptIn(ExperimentalTime::class)
+class ElectrumMiniWalletIntegrationTest : LightningTestSuite() {
+
+    private suspend fun CoroutineScope.connectToMainnetServer(): ElectrumClient {
+        val client =
+            ElectrumClient(TcpSocket.Builder(), this).apply { connect(ServerAddress("electrum.acinq.co", 50002, TcpSocket.TLS.UNSAFE_CERTIFICATES)) }
+
+        client.connectionState.first { it is Connection.CLOSED }
+        client.connectionState.first { it is Connection.ESTABLISHING }
+        client.connectionState.first { it is Connection.ESTABLISHED }
+
+        return client
+    }
+
+    @Test
+    fun `connect to an electrumx mainnet server`() = runSuspendTest(timeout = Duration.seconds(15)) { connectToMainnetServer().stop() }
+
+    @Test
+    fun `single address with no utxos`() = runSuspendTest(timeout = Duration.seconds(15)) {
+        val client = connectToMainnetServer()
+        val wallet = ElectrumMiniWallet(Block.LivenetGenesisBlock.hash, client, this)
+        wallet.addWatchOnlyAddress("bc1qyjmhaptq78vh5j7tnzu7ujayd8sftjahphxppz")
+
+        val walletState = wallet.walletStateFlow
+            .filter { it.addresses.size == 1 }
+            .first()
+
+        assertEquals(0, walletState.utxos.size)
+        assertEquals(0.sat, walletState.balance)
+
+        client.stop()
+    }
+
+    @Test
+    fun `single address with existing utxos`() = runSuspendTest(timeout = Duration.seconds(15)) {
+        val client = connectToMainnetServer()
+        val wallet = ElectrumMiniWallet(Block.LivenetGenesisBlock.hash, client, this)
+        wallet.addWatchOnlyAddress("14xb2HATmkBzrHf4CR2hZczEtjYpTh92d2")
+
+        val walletState = wallet.walletStateFlow
+            .filter { it.addresses.size == 1 }
+            .first()
+
+        assertEquals(6, walletState.utxos.size)
+        assertEquals(30_000_000.sat, walletState.balance)
+
+        client.stop()
+    }
+
+    @Test
+    fun `multiple addresses`() = runSuspendTest(timeout = Duration.seconds(15)) {
+        val client = connectToMainnetServer()
+        val wallet = ElectrumMiniWallet(Block.LivenetGenesisBlock.hash, client, this)
+        wallet.addAddress("16MmJT8VqW465GEyckWae547jKVfMB14P8")
+        wallet.addAddress("14xb2HATmkBzrHf4CR2hZczEtjYpTh92d2", randomKey())
+        wallet.addAddress("19b3QuFuYSSYPoLt1AxVQmG3LifgSSPNyA", randomKey())
+
+        val walletState = wallet.walletStateFlow
+            .filter { it.parentTxs.size == 11 }
+            .first()
+
+        assertEquals(4 + 6 + 1, walletState.utxos.size)
+        assertEquals(72_000_000.sat + 30_000_000.sat + 5_000_000.sat, walletState.balance)
+        assertEquals(7, walletState.spendable().size)
+        assertContains(
```

Comment:
We should probably verify that the previousTx's output correctly sends the right amount to the address we expect, this would ensure the `txid` / `tx_hash` confusion has been fixed. Overall this test should verify more things to ensure that we fetch and return the right data.

Reply:
I tried to improve with fbb1eed but I'm not convinced. I think you would prefer (1) a "self-referential test" where e.g. we verify that the output of a given tx indeed corresponds to an address defined in the test, while instead I went with (2) a "test vector" which compares to expected result to data taken from the blockchain.

For this particular test I think (2) is better than (1), because even if (1) is more self-explanatory, it is more verbose and ultimately only verifies a subset of (2).

Tbh I think the best would be to just build a complete `WalletState` and just do an `assertEquals`.

Reply:
You're right, the test vector is probably good enough, I verified it by looking at the blockchain, looks good!